// std-queue/dist/src/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const subqueue_1 = require("./subqueue");
class Queue {
    constructor() {
        this.top = new subqueue_1.default();
        this.bottom = this.top;
        this._size = 0;
    }
    get length() {
        return this._size;
    }
    push(...elems) {
        for (const elem of elems) {
            this.bottom.enqueue(elem);
            if (this.bottom.full()) {
                this.bottom = this.bottom.next = new subqueue_1.default();
            }
        }
        this._size += elems.length;
    }
    shift() {
        if (this._size === 0) {
            return undefined;
        }
        const val = this.top.dequeue();
        this._size--;
        if (this.top.size === 0 && this.top.full()) {
            this.top = this.top.next;
        }
        return val;
    }
    peek() {
        return this.top.peek();
    }
    last() {
        return this.bottom.last();
    }
    clear() {
        this.bottom = this.top = new subqueue_1.default();
        this._size = 0;
    }
}
exports.default = Queue;
